### ISSUE TITLE
Cannot set core config (eg, dialect or templater) inline

### ISSUE THREAD

**Issue by nspcc-cm** [sqlfluff/sqlfluff#7114]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I have the following .sql file:

```sql
-- sqlfluff:dialect:tsql
-- sqlfluff:templater:placeholder
-- sqlfluff:templater:placeholder:param_style:question_mark
-- sqlfluff:templater:placeholder:1:'2000-01-01'
-- sqlfluff:templater:placeholder:2:'2000-01-01'

select *
from dbo.my_table
where date_column between ? and ?;
```

Running a `sqlfluff parse` (or `fix`, etc) command I get a "no dialect was specified" message.

If I create a `.sqlfluff` file and move the dialect in there:

```ini
[sqlfluff]
dialect = tsql
```

...and remove the `sqlfluff:dialect:tsql` inline configuration from the .sql file and run the `sqlfluff parse` command again I get a "found unparsable section" message.

If I then move the `templater = placeholder` config from the .sql file to the .sqlfluff file and run the `sqlfluff parse` command, things work as expected.

This leads me to think that core config items (usually found under the [sqlfluff] section of a .sqlfluff file) are not being parsed and set properly.

### Expected Behaviour

It should be possible to parse/fix/etc the .sql file with inline configuration without the need for a .sqlfluff file.

### Observed Behaviour

The inline configuration would only work once the "core" configuration items were moved from inline to a .sqlfluff file.

### How to reproduce

Follow the steps in the What Happened section.

### Dialect

tsql

### Version

sqlfluff, version 3.4.2
Python 3.11.9

### Configuration

See What Happened.

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by chiruu12** [comment]
Reproduced on current main (4.3.0). The dialect half has a clear cause.

`Linter.load_raw_file_and_config` calls `root_config.make_child_from_path(fname)`, which defaults to `require_dialect=True`. That raises before the file is opened, so the `-- sqlfluff:dialect:tsql` line never gets a chance to supply the dialect. The inline scan (`process_raw_file_for_config`) runs four lines later, and `render_string` already re-checks with `verify_dialect_specified()` once the file has been read, with a comment saying that is where the check belongs.

Passing `require_dialect=False` at that call site fixes it. A file that sets a dialect nowhere still gets the same user error, just from the later check:

    $ sqlfluff parse t.sql     # with -- sqlfluff:dialect:tsql at the top
    [L:  1, P:  1]      |file:
    ...
    $ sqlfluff lint nodialect.sql
    User Error: No dialect was specified. ...

test/core and test/cli pass with that change (1231 passed; the two plugin_test failures are the example plugin not being installed and fail on a clean tree too).

The templater half is separate and I do not think it belongs in the same change. `Linter` picks its templater once at construction from the root config, so an inline `-- sqlfluff:templater:placeholder` cannot swap it, and `render_string` logs the "Attempt to set templater ... failed" warning for exactly that. Making the templater per-file is a bigger design call.

Happy to put up a PR for the dialect fix with a test. Say if you would rather the templater side went in the same PR.
